storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncProcessEvent.java

```diff
@@ -0,0 +1,672 @@
+                state = State.DISALLOWED;
+            } else if (supervisorData.getDeadWorkers().contains(workerId)) {
+                LOG.info("Worker Process {} has died", workerId);
+                state = State.TIMED_OUT;
+            } else if (SupervisorUtils.isWorkerHbTimedOut(now, whb, conf)) {
+                state = State.TIMED_OUT;
+            } else {
+                state = State.VALID;
+            }
+            LOG.debug("Worker:{} state:{} WorkerHeartbeat:{} at supervisor time-secs {}", workerId, state, whb, now);
+            workerIdHbstate.put(workerId, new StateHeartbeat(state, whb));
+        }
+        return workerIdHbstate;
+    }
+
+    protected boolean matchesAssignment(LSWorkerHeartbeat whb, Map<Integer, LocalAssignment> assignedExecutors) {
+        LocalAssignment localAssignment = assignedExecutors.get(whb.get_port());
+        if (localAssignment == null || !localAssignment.get_topology_id().equals(whb.get_topology_id())) {
+            return false;
+        }
+        List<ExecutorInfo> executorInfos = new ArrayList<>();
+        executorInfos.addAll(whb.get_executors());
+        // remove SYSTEM_EXECUTOR_ID
+        executorInfos.remove(SYSTEM_EXECUTOR_INFO);
+        List<ExecutorInfo> localExecuorInfos = localAssignment.get_executors();
+
+        if (localExecuorInfos.size() != executorInfos.size())
+            return false;
+
+        for (ExecutorInfo executorInfo : localExecuorInfos){
+            if (!localExecuorInfos.contains(executorInfo))
+                return false;
+        }
+        return true;
+    }
+
+    /**
+     * launch a worker in local mode.
+     */
+    protected void launchWorker(SupervisorData supervisorData, String stormId, Long port, String workerId, WorkerResources resources) throws IOException {
+        // port this function after porting worker to java
+    }
+
+    protected String getWorkerClassPath(String stormJar, Map stormConf) {
+        List<String> topoClasspath = new ArrayList<>();
+        Object object = stormConf.get(Config.TOPOLOGY_CLASSPATH);
+
+        if (object instanceof List) {
+            topoClasspath.addAll((List<String>) object);
+        } else if (object instanceof String){
+            topoClasspath.add((String)object);
+        }else {
+            //ignore
+        }
+        String classPath = Utils.workerClasspath();
+        String classAddPath = Utils.addToClasspath(classPath, Arrays.asList(stormJar));
+        return Utils.addToClasspath(classAddPath, topoClasspath);
+    }
+
+    /**
+     * "Generates runtime childopts by replacing keys with topology-id, worker-id, port, mem-onheap"
+     * 
+     * @param value
+     * @param workerId
+     * @param stormId
+     * @param port
+     * @param memOnheap
+     */
+    public List<String> substituteChildopts(Object value, String workerId, String stormId, Long port, int memOnheap) {
+        List<String> rets = new ArrayList<>();
```

Comment:
clojure code would have returned null if value is null. Also default type was assumed to be string.